In [1]:
import pandas as pd

## 1 задача

- Считать данные с помощью pandas

In [4]:
df = pd.read_csv('kc_house_data.csv')

- Вывести на экран первые 5 строк. Посмотреть на описание признаков и на их содержание.

In [10]:
df.head(6) # 5 строк таблицы

id             date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000   221900.0         3       1.00         1180   
1  6414100192  20141209T000000   538000.0         3       2.25         2570   
2  5631500400  20150225T000000   180000.0         2       1.00          770   
3  2487200875  20141209T000000   604000.0         4       3.00         1960   
4  1954400510  20150218T000000   510000.0         3       2.00         1680   
5  7237550310  20140512T000000  1225000.0         4       4.50         5420   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   
5    101930     1.0           0     0  ...     11        3890           1530   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   
5      2001             0    98053  47.6561 -122.005           4760   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  
5      101930  

[6 rows x 21 columns]

## 2 задача

- Проведите первичный анализ данных
- Посчитайте основные статистики по всем признакам и поизучайте их

In [9]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

- Изучите типы данных
- Найдите количество пропущенных ячеек в данных

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [23]:
df.describe(include=['object'])

date
count             21613
unique              372
top     20140623T000000
freq                142

Вывод: 
- всего 21613 лотов
- средняя цена ~540 тыс
- выставлены и студии и большие дома, исходя из данных по кол-ву спален, ванных комнат и площади
- дома с 1971 года постройки, до 2015 (дата данных)
- кол-во дней с единичной продажей - 372, 
- самое большее кол-во сделок: 142, на дату: 23 июня 2014

## 3 задача

### 3.1
В каком диапазоне изменяются стоимости недвижимости?

In [11]:
df['price'].min(), df['price'].max()

(75000.0, 7700000.0)

### 3.2
Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [14]:
living = df['sqft_living'].mean()
lot = df['sqft_lot'].mean()
(living/lot)*100

13.76781757959227

### 3.3
Как много домов с разными этажами в данных?

In [37]:
df[~(df['floors'] == 1)].shape[0]

10933

### 3.4
Насколько хорошие состояния у домов в данных?

In [79]:
df.condition.describe()

count    21613.000000
mean         3.409430
std          0.650743
min          1.000000
25%          3.000000
50%          3.000000
75%          4.000000
max          5.000000
Name: condition, dtype: float64

### 3.5
Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [36]:
df.yr_built.min(), df.yr_built.max()

(1900, 2015)

## 4 задание

### 4.1
Сколько в среднем стоят дома, у которых 2 спальни?

In [43]:
df[df['bedrooms']==2]['price'].mean().round(0)

401373.0

### 4.2 
Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [46]:
df[df['price']>600000]['sqft_lot'].mean().round(0)

20443.0

### 4.3 
Как много домов коснулся ремонт?

In [57]:
df[~(df['yr_renovated']==0)].shape[0]

914

### 4.4 
Насколько в среднем стоимость домов с оценкой grade выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [62]:
grade_10 = df[df['grade']==10]['price'].mean()
grade_4 = df[df.grade == 4].price.mean()
diff = (grade_10/grade_4)*100
diff.round(0)

500.0

## 5 задание

### 5.1 
Выберите дом клиенту. 
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [69]:
df[(df.waterfront == 1) & (df.bathrooms > 3) & (df.sqft_basement > 0)].shape[0]

36

### 5.2 
Выберите дом клиенту.
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [74]:
df[(df.view == 1)|(df.waterfront==1)&(df.grade >=10)&(df.yr_built>1980)]['price'].min(), df[(df.view == 1)|(df.waterfront==1)&(df.grade >=10)&(df.yr_built>1980)]['price'].max()

(217000.0, 5300000.0)

### 5.3 
Выберите дом клиенту. Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [77]:
df[(df.sqft_basement==0)&(df.floors==2)&(df.price<150000)]['grade'].mean()

7.0